Widgets should use autocompletes when there are many records

dsenalik opened the pull request:
# New Feature

### Closes #2074 

## Description
The current state of most core chado field widgets uses a dropdown select to add new records.
This is a problem when there are many records to choose from, as it can slow down page display tremendously. Currently on my migrated site, I cannot add or edit some content types through the UI because of this.

The solution proposed here is to automatically switch over to an autocomplete field when the number of records is large.
I arbitrarily chose 50 as the limit. This can be configured both globally and on a per-field basis.

Implementation of so many autocompletes was by creating a "Generic" chado autocomplete, that can be used for any chado table where the value comes from a single column. This means it can't be used for organism, so that field remains unchanged.

~~The existing project autocomplete has been removed and replaced with the generic version.~~
Tests are now kernel instead of functional, so a bit of an improvement there.

For tables with a type_id column, the type is included in the select menus in square brackets. e.g. "PI 123456 [Germplasm Accession]" for the stock field.

All instances of "-- Select --" have been changed to "- Select -" to match what Drupal uses.

## Testing?
Since most chado linking fields are updated, manual testing will require creating at least one of a content type for each core table currently supported with a field.
You can toggle select vs. autocomplete by setting the global limit at `/admin/tripal/config/widget_settings` 

As a select:
![PR2075 2](https://github.com/user-attachments/assets/41cc06a0-8952-4933-be0c-4dc1d77952e1)

As an autocomplete:
![PR2075 3](https://github.com/user-attachments/assets/1d2bcff3-96ca-482a-9baf-bf3a031b127a)


## ToDo
 - [x] Configuration option for the limit
![PR2075 1](https://github.com/user-attachments/assets/4705f3a8-7918-441d-9c16-480af9dbdf59)

dsenalik committed:
first draft of generic autocomplete controller

dsenalik committed:
a function to autoselect select vs. autocomplete

dsenalik committed:
pub widget autocomplete

dsenalik committed:
pub widget autocomplete

dsenalik committed:
update autocomplete

dsenalik committed:
rename to genericSelectElement

dsenalik committed:
update pub widget

dsenalik committed:
analysis widget autocomplete

dsenalik committed:
kernel tests for generic autocomplete

dsenalik committed:
fix path of generic autocomplete

dsenalik committed:
white space

dsenalik committed:
use @group Autocomplete for all autocompletes

dsenalik committed:
project widget autocomplete

dsenalik committed:
move functional project autocomplete tests to kernel generic autocomplete

dsenalik committed:
validate and massage when using autocomplete

dsenalik committed:
arraydesign widget autocomplete

dsenalik committed:
assay widget autocomplete

dsenalik committed:
biomaterial widget autocomplete

dsenalik committed:
contact widget autocomplete

dsenalik committed:
feature widget autocomplete

dsenalik committed:
featuremap widget autocomplete

dsenalik committed:
protocol widget autocomplete

dsenalik committed:
stock widget autocomplete

dsenalik committed:
study widget autocomplete

dsenalik committed:
genericSelectMassageFormValues

dsenalik committed:
quoting required

dsenalik committed:
include type when present

dsenalik committed:
comment refinements; default to 10

dsenalik committed:
fix 'type' on contact query

dsenalik committed:
add 'type' to stock

dsenalik committed:
add 'type' to pub

dsenalik committed:
Merge branch '4.x' into tv4g10-issue2074-more-autocompletes-for-fields

dsenalik committed:
change "-- Select --" to "- Select -", same as Drupal uses

dsenalik committed:
remove project autocomplete

dsenalik committed:
add settings form for select vs. autocomplete

laceysanderson reviewed (changes requested):
## Project autocomplete controller

You shouldn't delete the project autocomplete 😬 I have extension modules that are using it 🙈 instead you should keep it around but inherit from the Generic Autocomplete controller to reduce duplication.

## ChadoGenericAutocompleteController Query

In tripal_chado/src/Controller/ChadoGenericAutocompleteController.php you use concatenation to build the query. We should try to move away from this approach and use the query builder instead since its more readable IMO and less buggy in general.

```php
$query = $connection->select('BT', '1:' . $base_table);
$query->addField('BT', $pkey_id, 'pkey');
$query->addField('BT', $column_name, 'value');
$query->condition('LOWER(BT.' . $column_name .')', $keyword, 'LIKE');
$query->orderBy('BT.' . $column_name, 'ASC');
$query->range(0, $count);

if ($type_id > 0) {
  if ($property_table and ($property_table != $base_table)) {
    $query->join('1:'.$property_table, 'PT', 'BT.' . $pkey_id . ' = PT.' . $pkey_id);
    $query->condition('PT.' . $type_column, $type_id);
  }
  else {
    $query->condition('BT.' . $type_column, $type_id);
  }
}
```

## Restrict by type or include

> For tables with a type_id column, the type is included in the select menus in square brackets. e.g. "PI 123456 [Germplasm Accession]" for the stock field.

This part doesn't seem to be implemented 🤔 There is no adding of the cvterm name to the query and instead you just filter to the type. However, we do need to support cross-type autocompletes where we don't filter but rather add into the result. I would do this if the type_id supplied is `0`.

## Possible query mismatch

In `tripal_chado/src/TripalField/ChadoWidgetBase.php` the genericSelectElement() method allows widgets to pass in a query for select but if there are too many results it will use the controlled to create an autocomplete query. This could result in mismatched queries. Instead I would add a static `getQuery` method to the autocomplete controller that has an option indicating if the query should be limited or not. That way, `getQuery` can be used by both the `handleAutocomplete()` and `genericSelectElement()`.

## Configuration

> The solution proposed here is to automatically switch over to an autocomplete field when the number of records is large.
I arbitrarily chose 50 as the limit.

I see you added a stand alone configuration form for the limit but 

1. I think it should just have been added to the existing entity.tripal_entity_settings form.
2. I think that form sets the global default but it should also be set in the widget settings per widget. 

Also, I don't think we should rely only on the auto-switch. Count queries alone on my site will cause a page time out 🙈 As such, the widget should also have a setting for when to use select or autocomplete. Something like

> Behaviour for widget type:
> 🔘 Always Select List but limit to maximum results
> 🔘 Always Autocomplete
> 🔘 Select list if less than maximum results and autocomplete otherwise.

Since it is now configurable per content type, I would use a default of "Select list if less than maximum results and autocomplete otherwise" for the content types we know are usually small lists (like organism) and "Always Autocomplete" on most.

dsenalik committed:
Merge branch '4.x' into tv4g10-issue2074-more-autocompletes-for-fields

dsenalik committed:
restore the project autocomplete

dsenalik committed:
switch to query builder

dsenalik committed:
implement getQuery()

dsenalik committed:
move global widget setting to tripal entity form

dsenalik committed:
Implement individual widget settings

dsenalik committed:
Match Drupal, zero for autocomplete limit is unlimited

dsenalik commented:
> ## Project autocomplete controller
> 
> ...You shouldn't delete the project autocomplete 😬 ...

Sorry, it's back! Nothing in core Tripal was using it, but I didn't think about other modules 😬.

> ## ChadoGenericAutocompleteController Query
> 
> ...We should try to move away from this approach and use the query builder instead...

I have done so.
 
> ## Restrict by type or include
> 
> > For tables with a type_id column, the type is included in the select menus in square brackets. e.g. "PI 123456 [Germplasm Accession]" for the stock field.
> 
> This part doesn't seem to be implemented 🤔...

This had been implemented only for the select options, but not for the autocomplete. This is now implemented for both select and autocomplete **in cases where `type_id` is in the base table**. It is not done for cases where the bundle is defined in the property table, due to the complexity of doing this, you need to know ahead of time which terms to query for, and this seemed to be too much to do inside an autocomplete, which needs to be fast.

> ## Possible query mismatch
> 
> ...Instead I would add a static `getQuery` method to the autocomplete controller that has an option indicating if the query should be limited or not. That way, `getQuery` can be used by both the `handleAutocomplete()` and `genericSelectElement()`...

This is done, and I think makes the code clearer too.

> ## Configuration
> 
> 1. I think it should just have been added to the existing entity.tripal_entity_settings form.

Moved there now.
![2075 0111 1](https://github.com/user-attachments/assets/5713a637-a7dd-4dca-868d-19b7cff39a2b)

> 2. I think that form sets the global default but it should also be set in the widget settings per widget.

That was a job, but I did it. I copied Drupal almost verbatim here and because of that included several settings.
If set for the field, use that. If not, use the global setting. If that's not set, default to 50.
I copied the "Authored by" autocomplete settings which appear as
![PR2075 1012 1](https://github.com/user-attachments/assets/78eed2e6-dada-4137-bf33-f7f00e058fed)
![PR2075 1012 2](https://github.com/user-attachments/assets/95f30762-9787-4d1e-ae06-fa49bb483bb8)

You can also set the number of options returned by the autocomplete now, change the field size, and specify a placeholder if you want.
![2075 0111 2](https://github.com/user-attachments/assets/1bc5be8b-d964-4fbb-9ddf-e45a0be4d261)
![2075 0111 3](https://github.com/user-attachments/assets/9f949adb-bdcd-45b4-aa2d-8a2097ab8eb1)


> Also, I don't think we should rely only on the auto-switch. Count queries alone on my site will cause a page time out 🙈 As such, the widget should also have a setting for when to use select or autocomplete. Something like
> 
> > Behaviour for widget type:
> > 🔘 Always Select List but limit to maximum results
> > 🔘 Always Autocomplete
> > 🔘 Select list if less than maximum results and autocomplete otherwise.
> 
> Since it is now configurable per content type, I would use a default of "Select list if less than maximum results and autocomplete otherwise" for the content types we know are usually small lists (like organism) and "Always Autocomplete" on most.

We should discuss this one, I can't see a reason to always have a select list but not return everything. It seems it would confuse people if this happened.
With the current implementation, you set zero for the limit to force always autocomplete.

Another point, the count is performed on a subquery, after the LIMIT is applied, so this is much more efficient than counting every record just to see if it is over the limit. So I am curious if this works just fine in your case even with millions of records. And regardless, if you set the limit to zero, the count will not even be performed.
![PR2075 1012 3](https://github.com/user-attachments/assets/ed8c2446-94bf-48aa-9185-32c475251214)

dsenalik committed:
default for field size

dsenalik committed:
fix schema errors (I forgot again)

dsenalik committed:
don't perform count query if select_limit is zero

dsenalik committed:
corrections to count logic

dsenalik committed:
Merge branch '4.x' into tv4g10-issue2074-more-autocompletes-for-fields

laceysanderson reviewed (commented):
So I'm a bit confused. I've built a fresh docker image container on this branch many times at this point, confirmed that your changes are in the container, cleared caches, etc and I'm still finding that most of the fields I expect to have the widget settings do not. 😕 

Do you see the widget settings cog consistently on these fields?

The following screenshots show that I can find the one analysis linker field with the settings but the contact and project linkers don't have widget settings? Also, shouldn't there be more analysis linkers? Why isn't there one on gene pages?

### Contact
<img width="1405" alt="Screenshot 2025-01-16 at 4 52 09 PM" src="https://github.com/user-attachments/assets/2e41d2f8-0323-4725-9255-4029ae1f63fb" />

### Project
<img width="1404" alt="Screenshot 2025-01-16 at 4 52 55 PM" src="https://github.com/user-attachments/assets/7e74d2d6-a71c-48fb-bdf6-97c98f09744b" />

### Gene
<img width="1281" alt="Screenshot 2025-01-16 at 4 54 29 PM" src="https://github.com/user-attachments/assets/be056e70-5fb9-4438-a776-c4b3bacee9ee" />

dsenalik committed:
Merge branch '4.x' into tv4g10-issue2074-more-autocompletes-for-fields

dsenalik committed:
refinement of comments

dsenalik committed:
add settings to remaining fields

dsenalik committed:
Merge branch '4.x' into tv4g10-issue2074-more-autocompletes-for-fields

dsenalik commented:
This was a case of slovenly testing on my part. I forgot to add settings to most of the fields that needed them.
Now gene for example has the missing settings:
![2075 gene](https://github.com/user-attachments/assets/36284dad-2817-4207-9862-cde238183108)
Contact settings shown in this screenshot have been edited to all non-default values

I will remove draft status only after properly testing on my end 🙄

dsenalik committed:
join needs to be leftJoin

dsenalik committed:
no brackets if type is not present

dsenalik committed:
call parent functions

dsenalik committed:
widget base minor edit

dsenalik committed:
Merge branch '4.x' into tv4g10-issue2074-more-autocompletes-for-fields

dsenalik committed:
placeholder is a string

dsenalik committed:
simplify placeholder summary

dsenalik committed:
change back to match Drupal style

dsenalik committed:
fix query for autocomplete default value

dsenalik committed:
Merge branch '4.x' into tv4g10-issue2074-more-autocompletes-for-fields

dsenalik committed:
Merge branch '4.x' into tv4g10-issue2074-more-autocompletes-for-fields

dsenalik committed:
Merge branch '4.x' into tv4g10-issue2074-more-autocompletes-for-fields

dsenalik committed:
set global default if not defined

laceysanderson reviewed (approved):
Cases checked:
A) Has indicated settings for all fields I would expect.
B) Shows default max records of 50 on first view
C) Shows updated global default when field-specific settings have not been updated

Regarding A, we might want to enable these settings for the "Database Reference Annotations", "Synonyms" and "Organism" too at some point --talk to Doug about this already and we will make separate issues. I was also happy to see the settings for discovered fields such as the "Sequence Feature" field on analysis.

|                             | A | B | C |
|-----------------------------|---|---|---|
| Analysis                    | ✅ | ✅ | ✅ |
| Contact                     | ✅ | ✅ | ✅ |
| Organism                    | ✅ | ✅ | ✅ |
| Project                     | ✅ | ✅ | ✅ |
| Protocol                    | ✅ | ✅ | ✅ |
| Publication                 | ✅ | ✅ | ✅ |
| Study                       | ✅ | ✅ | ✅ |
| Breeding Cross              | ✅ | ✅ | ✅ |
| Germplasm Accession         | ✅ | ✅ | ✅ |
| Germplasm Variety           | ✅ | ✅ | ✅ |
| Recombinant Inbred line     | ✅ | ✅ | ✅ |
| DNA Library                 | ✅ | ✅ | ✅ |
| Gene                        | ✅ | ✅ | ✅ |
| Genome Annotation           | ✅ | ✅ | ✅ |
| Genome Assembly             | ✅ | ✅ | ✅ |
| Genome Project              | ✅ | ✅ | ✅ |
| mRNA                        | ✅ | ✅ | ✅ |
| Phylogenetic Tree           | ✅ | ✅ | ✅ |
| Physical Map                | ✅ | ✅ | ✅ |
| Genetic Map                 | ✅ | ✅ | ✅ |
| Genetic Marker              | ✅ | ✅ | ✅ |
| Heritable Phenotypic Marker | ✅ | ✅ | ✅ |
| QTL                         | ✅ | ✅ | ✅ |
| Sequence Variant            | ✅ | ✅ | ✅ |
| Array Design                | ✅ | ✅ | ✅ |
| Assay                       | ✅ | ✅ | ✅ |
| Biological Sample           | ✅ | ✅ | ✅ |

Cases checked on a few fields but not all like above:
- ✅ Field settings are saved and override the global default
- ✅ Widget does switch between select and autocomplete when too many records and if configured to do so
- ✅ Widget does not switch regardless of number of records if configured not to.
- ✅ If widget is configured to switch at a higher number of records then global, ensure the local is respected

dsenalik commented:
I created issue #2106 to acknowledge that the organism field needs an autocomplete
And issue #2107 to add similar settings for dbxref and synonym fields
